File: paradigms/factorial_fibonacci.py
"""
Recursion vs Iteration in Python
"""

# --- Factorial ---

def factorial_recursive(n: int) -> int:
    if n <= 1:
        return 1
    return n * factorial_recursive(n - 1)

def factorial_iterative(n: int) -> int:
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result

# --- Fibonacci ---

def fibonacci_recursive(n: int) -> int:
    if n <= 1:
        return n
    return fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2)

def fibonacci_iterative(n: int) -> int:
    if n <= 1:
        return n
    prev, curr = 0, 1
    for _ in range(2, n + 1):
        prev, curr = curr, prev + curr
    return curr

if __name__ == "__main__":
    n_factorial = 5
    n_fibonacci = 10

    print(f"--- Factorial of {n_factorial} ---")
    print(f"Recursive: {factorial_recursive(n_factorial)}")
    print(f"Iterative: {factorial_iterative(n_factorial)}")

    print(f"\n--- Fibonacci of {n_fibonacci} ---")
    print(f"Recursive: {fibonacci_recursive(n_fibonacci)}")
    print(f"Iterative: {fibonacci_iterative(n_fibonacci)}")

File: paradigms/factorial_fibonacci.go
package main

import "fmt"

/**
 * Recursion vs Iteration in Go
 */

// --- Factorial ---

func factorialRecursive(n int) int {
	if n <= 1 {
		return 1
	}
	return n * factorialRecursive(n-1)
}

func factorialIterative(n int) int {
	result := 1
	for i := 2; i <= n; i++ {
		result *= i
	}
	return result
}

// --- Fibonacci ---

func fibonacciRecursive(n int) int {
	if n <= 1 {
		return n
	}
	return fibonacciRecursive(n-1) + fibonacciRecursive(n-2)
}

func fibonacciIterative(n int) int {
	if n <= 1 {
		return n
	}
	prev, curr := 0, 1
	for i := 2; i <= n; i++ {
		prev, curr = curr, prev+curr
	}
	return curr
}

func main() {
	nFactorial := 5
	nFibonacci := 10

	fmt.Printf("--- Factorial of %d ---\n", nFactorial)
	fmt.Printf("Recursive: %d\n", factorialRecursive(nFactorial))
	fmt.Printf("Iterative: %d\n", factorialIterative(nFactorial))

	fmt.Printf("\n--- Fibonacci of %d ---\n", nFibonacci)
	fmt.Printf("Recursive: %d\n", fibonacciRecursive(nFibonacci))
	fmt.Printf("Iterative: %d\n", fibonacciIterative(nFibonacci))
}
